[JSC] Optimize `Array#lastIndexOf` in DFG/FTL

sosukesuzuki opened the pull request:
#### bc818abb8d4b41975712c2c6ec66bf7500753351
<pre>
[JSC] Optimize `Array#lastIndexOf` in DFG/FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=300170">https://bugs.webkit.org/show_bug.cgi?id=300170</a>

Reviewed by Yusuke Suzuki.

This patch changes to optimize `Array#lastIndexOf` in DFG/FTL.

                                                  TipOfTree                  Patched

array-prototype-lastIndexOf-string-16           0.7393+-0.0963            0.6353+-0.0238          might be 1.1637x faster
array-prototype-lastIndexOf-double-from-contiguous
                                                0.7476+-0.3292            0.6312+-0.0984          might be 1.1845x faster
array-prototype-lastIndexOf-int32               3.8388+-0.8260     ^      2.3827+-0.3433        ^ definitely 1.6111x faster
array-prototype-lastIndexOf-bigint              9.9618+-0.7504     ^      8.3997+-0.4838        ^ definitely 1.1860x faster
array-prototype-lastIndexOf-contiguous          5.4971+-1.0560     ^      2.1412+-0.1266        ^ definitely 2.5673x faster
array-prototype-lastIndexOf-string-const        3.7607+-0.1203     ^      2.9016+-0.1071        ^ definitely 1.2961x faster
array-prototype-lastIndexOf-untyped-int32
                                               80.1997+-0.6415     ^     42.8832+-0.8174        ^ definitely 1.8702x faster
array-prototype-lastIndexOf-int32-from-contiguous
                                                0.6262+-0.1345            0.6172+-0.0809          might be 1.0146x faster
array-prototype-lastIndexOf-double              2.3190+-0.0583     ^      1.3878+-0.0684        ^ definitely 1.6711x faster
array-prototype-lastIndexOf-string              0.7423+-0.1096            0.6556+-0.1410          might be 1.1323x faster

Tests: JSTests/microbenchmarks/array-prototype-lastIndexOf-bigint.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-contiguous.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-double-from-contiguous.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-double.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-int32-from-contiguous.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-int32.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-string-16.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-string-const.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-string.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-untyped-int32.js
       JSTests/stress/array-prototype-lastIndexOf-doublerepuse.js
       JSTests/stress/array-prototype-lastIndexOf-int32-from-contiguous.js
       JSTests/stress/array-prototype-lastIndexOf-int32use.js
       JSTests/stress/array-prototype-lastIndexOf-objectuse.js
       JSTests/stress/array-prototype-lastIndexOf-stringuse.js
       JSTests/stress/array-prototype-lastIndexOf-symboluse.js
       JSTests/stress/array-prototype-lastIndexOf-untypeduse-contiguous.js
       JSTests/stress/array-prototype-lastIndexOf-untypeduse-int32.js
* JSTests/microbenchmarks/array-prototype-lastIndexOf-bigint.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-contiguous.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-double-from-contiguous.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-double.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-int32-from-contiguous.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-int32.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-lastIndexOf-string-16.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-lastIndexOf-string-const.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-lastIndexOf-string.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-lastIndexOf-untyped-int32.js: Added.
(test):
* JSTests/stress/array-prototype-lastIndexOf-doublerepuse.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-int32-from-contiguous.js: Added.
(shouldBe):
* JSTests/stress/array-prototype-lastIndexOf-int32use.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-objectuse.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-stringuse.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-symboluse.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-untypeduse-contiguous.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-untypeduse-int32.js: Added.
(sameValue):
(test):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
(JSC::DFG::FixupPhase::fixupArrayIndexOfOrArrayIncludes):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::arrayLastIndexOfString):
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOfOrArrayIncludes):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
* Source/JavaScriptCore/runtime/Intrinsic.h:

Canonical link: <a href="https://commits.webkit.org/302157@main">https://commits.webkit.org/302157@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7adce57680f571ce50311aed961393aff8b2c0a9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124631 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44303 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35037 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131471 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76561 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52873 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94815 "Found 3 new test failures: imported/w3c/web-platform-tests/event-timing/crossiframe.html imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub.html imported/w3c/web-platform-tests/webrtc/simulcast/basic.https.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62873 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127585 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35888 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111459 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75387 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29615 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74952 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116736 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105644 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29845 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134138 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123151 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51479 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39300 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103290 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51891 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107681 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103068 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26987 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48439 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26703 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48420 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51353 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57150 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156280 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50748 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39152 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54102 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52442 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -4241,6 +4241,168 @@ JSC_DEFINE_JIT_OPERATION(operationCopyOnWriteArrayIndexOfString, UCPUStrictInt32
+        OPERATION_RETURN(scope, toUCPUStrictInt32(-1));
+
+    EncodedJSValue encodedSearchElement = JSValue::encode(jsNumber(int32Value));
+    for (; index >= 0; --index) {
```
you should be able to get a lot better performance on this when the input is long if you use simd

sosukesuzuki replied:
But there isn't backwards version of `WTF::find64` in WebKit, so I want to make it out of scope in this PR

sosukesuzuki committed:
[JSC] Optimize `Array#lastIndexOf` in DFG/FTL
https://bugs.webkit.org/show_bug.cgi?id=300170

Reviewed by Yusuke Suzuki.

This patch changes to optimize `Array#lastIndexOf` in DFG/FTL.

                                                  TipOfTree                  Patched

array-prototype-lastIndexOf-string-16           0.7393+-0.0963            0.6353+-0.0238          might be 1.1637x faster
array-prototype-lastIndexOf-double-from-contiguous
                                                0.7476+-0.3292            0.6312+-0.0984          might be 1.1845x faster
array-prototype-lastIndexOf-int32               3.8388+-0.8260     ^      2.3827+-0.3433        ^ definitely 1.6111x faster
array-prototype-lastIndexOf-bigint              9.9618+-0.7504     ^      8.3997+-0.4838        ^ definitely 1.1860x faster
array-prototype-lastIndexOf-contiguous          5.4971+-1.0560     ^      2.1412+-0.1266        ^ definitely 2.5673x faster
array-prototype-lastIndexOf-string-const        3.7607+-0.1203     ^      2.9016+-0.1071        ^ definitely 1.2961x faster
array-prototype-lastIndexOf-untyped-int32
                                               80.1997+-0.6415     ^     42.8832+-0.8174        ^ definitely 1.8702x faster
array-prototype-lastIndexOf-int32-from-contiguous
                                                0.6262+-0.1345            0.6172+-0.0809          might be 1.0146x faster
array-prototype-lastIndexOf-double              2.3190+-0.0583     ^      1.3878+-0.0684        ^ definitely 1.6711x faster
array-prototype-lastIndexOf-string              0.7423+-0.1096            0.6556+-0.1410          might be 1.1323x faster

Tests: JSTests/microbenchmarks/array-prototype-lastIndexOf-bigint.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-contiguous.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-double-from-contiguous.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-double.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-int32-from-contiguous.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-int32.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-string-16.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-string-const.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-string.js
       JSTests/microbenchmarks/array-prototype-lastIndexOf-untyped-int32.js
       JSTests/stress/array-prototype-lastIndexOf-doublerepuse.js
       JSTests/stress/array-prototype-lastIndexOf-int32-from-contiguous.js
       JSTests/stress/array-prototype-lastIndexOf-int32use.js
       JSTests/stress/array-prototype-lastIndexOf-objectuse.js
       JSTests/stress/array-prototype-lastIndexOf-stringuse.js
       JSTests/stress/array-prototype-lastIndexOf-symboluse.js
       JSTests/stress/array-prototype-lastIndexOf-untypeduse-contiguous.js
       JSTests/stress/array-prototype-lastIndexOf-untypeduse-int32.js
* JSTests/microbenchmarks/array-prototype-lastIndexOf-bigint.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-contiguous.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-double-from-contiguous.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-double.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-int32-from-contiguous.js: Added.
* JSTests/microbenchmarks/array-prototype-lastIndexOf-int32.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-lastIndexOf-string-16.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-lastIndexOf-string-const.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-lastIndexOf-string.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-lastIndexOf-untyped-int32.js: Added.
(test):
* JSTests/stress/array-prototype-lastIndexOf-doublerepuse.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-int32-from-contiguous.js: Added.
(shouldBe):
* JSTests/stress/array-prototype-lastIndexOf-int32use.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-objectuse.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-stringuse.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-symboluse.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-untypeduse-contiguous.js: Added.
(sameValue):
(test):
* JSTests/stress/array-prototype-lastIndexOf-untypeduse-int32.js: Added.
(sameValue):
(test):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
(JSC::DFG::FixupPhase::fixupArrayIndexOfOrArrayIncludes):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::arrayLastIndexOfString):
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOfOrArrayIncludes):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
* Source/JavaScriptCore/runtime/Intrinsic.h:

Canonical link: https://commits.webkit.org/302157@main
